game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+package games.strategy.triplea.ai.tree;
+
+import com.google.common.base.Preconditions;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.TerritoryEffect;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.attachments.TechAbilityAttachment;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import games.strategy.triplea.delegate.battle.TargetGroup;
+import games.strategy.triplea.delegate.battle.casualty.CasualtyUtil;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Value;
+import org.apache.commons.math3.distribution.BinomialDistribution;
+import org.triplea.java.PredicateBuilder;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.java.collections.IntegerMap;
+import org.triplea.util.Triple;
+import org.triplea.util.Tuple;
+
+public class BattleStep {
+
+  // enable this to keep children in memory for analysis
+  // watch out for running out of memory, though
+  private static final boolean DEBUG = false;
+
+  private static final int MAX_ROUNDS = 16;
+  private static final double IGNORE_BRANCH_PROBABILITY = 0.005;
+
+  enum Type {
+    AA_ATTACKER,
+    AA_DEFENDER,
+    SUB_ATTACKER,
+    SUB_DEFENDER,
+    ATTACKER,
+    DEFENDER;
+
+    // TODO - the order of subs is configurable (see MustFightBattle::addFightSteps)
+    Type nextType() {
+      switch (this) {
+        case AA_ATTACKER:
+          return AA_DEFENDER;
+        case AA_DEFENDER:
+          return SUB_ATTACKER;
+        case SUB_ATTACKER:
+          return SUB_DEFENDER;
+        case SUB_DEFENDER:
+          return ATTACKER;
+        case ATTACKER:
+          return DEFENDER;
+        case DEFENDER:
+          return AA_ATTACKER;
+        default:
+          Preconditions.checkArgument(false, "A type has been forgotten in swapSides.");
+          return AA_ATTACKER;
+      }
```

Comment:
Unfortunately, since the order of sub attacks is configurable, that way won't work.  Once I've finished getting subs to work, it won't be a simple switch statement.